<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Estimating the reproduction number from modelled incidence â€¢ ggoutbreak</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="â€image/svg+xmlâ€" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimating the reproduction number from modelled incidence">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggoutbreak</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ggoutbreak.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/covid-timeseries.html">England COVID-19 cases</a></li>
    <li><a class="dropdown-item" href="../articles/estimators-example.html">Simulation tests for growth rate estimators</a></li>
    <li><a class="dropdown-item" href="../articles/incidence-trends.html">Population comparisons and incidence</a></li>
    <li><a class="dropdown-item" href="../articles/infectivity-profile-discretisation.html">Infectivity profile discretisation</a></li>
    <li><a class="dropdown-item" href="../articles/right-censoring.html">Estimating the reproduction number from right censored data</a></li>
    <li><a class="dropdown-item" href="../articles/rt-from-incidence.html">Estimating the reproduction number from modelled incidence</a></li>
    <li><a class="dropdown-item" href="../articles/sampling-serial-interval.html">Sampling the infectivity profile from published serial interval estimates</a></li>
    <li><a class="dropdown-item" href="../articles/simulation-test-models.html">Simulations and test harnesses</a></li>
    <li><a class="dropdown-item" href="../articles/time-periods.html">Data wrangling and working with `ggoutbreak`</a></li>
    <li><a class="dropdown-item" href="../articles/variant-proportions.html">Multinomial proportions models for genomic variants</a></li>
    <li><a class="dropdown-item" href="../articles/weekly-incidence.html">Estimating the reproduction number from weekly data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ai4ci/ggoutbreak/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Estimating the reproduction number from modelled incidence</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ai4ci/ggoutbreak/blob/HEAD/vignettes/rt-from-incidence.Rmd" class="external-link"><code>vignettes/rt-from-incidence.Rmd</code></a></small>
      <div class="d-none name"><code>rt-from-incidence.Rmd</code></div>
    </div>

    
    
<p>Robert Challen <sup>1,2</sup>; Leon Danon <sup>1,2</sup>;</p>
<ol style="list-style-type: decimal">
<li>Engineering Mathematics, University of Bristol, Bristol, UK</li>
<li>AI4CI</li>
</ol>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>If we have estimated the incidence of a disease
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>t</mi></msub><annotation encoding="application/x-tex">I_t</annotation></semantics></math>
using a poisson rate using maximum likelihood estimators, the rate is
typically a log-normally distributed with parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Î¼</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ïƒ</mi><annotation encoding="application/x-tex">\sigma</annotation></semantics></math>.
Such a fitted model is shown below on a log1p scale, for the COVID-19
epidemic in England:</p>
<p><img src="rt-from-incidence_files/figure-html/unnamed-chunk-1-1.png" width="576"></p>
<p>It is appealing to use this modelled incidence estimate to calculate
an estimate of the reproduction number,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>.
Incidence models can be derived in a number of ways, they are easily
inspected for error and can be made tolerant of missing values and
outliers.</p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>To use a modelled estimate of incidence to predict
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
we need to propagate uncertainty in incidence into our
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
estimates. To calculate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
we can use the backwards-looking renewal equations which incorporate the
infectivity profile of the disease
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ï‰</mi><annotation encoding="application/x-tex">\omega</annotation></semantics></math>)
at a number of days after infection
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ï„</mi><annotation encoding="application/x-tex">\tau</annotation></semantics></math>):</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>I</mi><mi>t</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>âˆ¼</mo><mi>P</mi><mi>o</mi><mi>i</mi><mi>s</mi><mi>s</mi><mi>o</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î»</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î»</mi><mi>t</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>âˆ¼</mo><mi>L</mi><mi>o</mi><mi>g</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mi>a</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mi>t</mi></msub><mo>,</mo><msub><mi>Ïƒ</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>R</mi><mi>t</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><msub><mi>I</mi><mi>t</mi></msub><mrow><munder><mo>âˆ‘</mo><mi>Ï„</mi></munder><mrow><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><msub><mi>I</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub></mrow></mrow></mfrac></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
I_t &amp;\sim Poisson(\lambda_t) \\
\lambda_t &amp;\sim Lognormal(\mu_t,\sigma_t) \\
R_t &amp;= \frac{I_t}{\sum_{\tau}{\omega_{\tau}I_{t-\tau}}}
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>In expectation, this gives:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>R</mi><mi>t</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>â‰ˆ</mo><mfrac><msub><mi>Î»</mi><mi>t</mi></msub><mrow><munderover><mo>âˆ‘</mo><mrow><mi>Ï„</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><msub><mi>Î»</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub></mrow></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>âˆ¼</mo><mfrac><mrow><mtext mathvariant="normal">Lognormal</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mi>t</mi></msub><mo>,</mo><msub><mi>Ïƒ</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munderover><mo>âˆ‘</mo><mrow><mi>Ï„</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><mtext mathvariant="normal">Lognormal</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub><mo>+</mo><mo>log</mo><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><mo>,</mo><msub><mi>Ïƒ</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
R_t &amp;\approx \frac{\lambda_t}{\sum_{\tau=1}^{k} \omega_\tau \lambda_{t-\tau}} \\
&amp;\sim \frac{\text{Lognormal}(\mu_t, \sigma_t)}{\sum_{\tau=1}^{k} \text{Lognormal}\left( \mu_{t-\tau} + \log \omega_\tau, \sigma_{t-\tau} \right)}
\end{aligned}
\tag{1}
\end{eqnarray}
</annotation></semantics></math></p>
<p>It has been shown that the sum of correlated log-normal distributed
random variables can be approximated by another log-normal with
parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¼</mi><mi>Z</mi></msub><annotation encoding="application/x-tex">\mu_Z</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>Z</mi></msub><annotation encoding="application/x-tex">\sigma_Z</annotation></semantics></math>,
where the correlation between them is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Ï</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mtext mathvariant="normal">Corr</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>X</mi><mi>i</mi></msub><mo>,</mo><mo>log</mo><msub><mi>X</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\rho_{ij} = \text{Corr}(\log X_i, \log X_j)</annotation></semantics></math>.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>S</mi><mo>+</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>ğ”¼</mi><mrow><mo stretchy="true" form="prefix">[</mo><munder><mo>âˆ‘</mo><mi>i</mi></munder><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><munder><mo>âˆ‘</mo><mi>i</mi></munder><mi>ğ”¼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><munder><mo>âˆ‘</mo><mi>i</mi></munder><msup><mi>e</mi><mrow><msub><mi>Î¼</mi><mi>i</mi></msub><mo>+</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mi>i</mi><mn>2</mn></msubsup></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><msubsup><mi>S</mi><mo>+</mo><mn>2</mn></msubsup></mfrac><mspace width="0.167em"></mspace><munder><mo>âˆ‘</mo><mrow><mi>i</mi><mo>,</mo><mi>j</mi></mrow></munder><msub><mi>Ï</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>Ïƒ</mi><mi>i</mi></msub><msub><mi>Ïƒ</mi><mi>j</mi></msub><mi>ğ”¼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mi>ğ”¼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>X</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><msubsup><mi>S</mi><mo>+</mo><mn>2</mn></msubsup></mfrac><mspace width="0.167em"></mspace><munder><mo>âˆ‘</mo><mrow><mi>i</mi><mo>,</mo><mi>j</mi></mrow></munder><msub><mi>Ï</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>Ïƒ</mi><mi>i</mi></msub><msub><mi>Ïƒ</mi><mi>j</mi></msub><msup><mi>e</mi><mrow><msub><mi>Î¼</mi><mi>i</mi></msub><mo>+</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mi>i</mi><mn>2</mn></msubsup></mrow></msup><msup><mi>e</mi><mrow><msub><mi>Î¼</mi><mi>j</mi></msub><mo>+</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mi>j</mi><mn>2</mn></msubsup></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î¼</mi><mi>Z</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mo>log</mo><msub><mi>S</mi><mo>+</mo></msub><mo>âˆ’</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
    S_+ &amp;= \mathbb{E}\left[\sum_i X_i \right] = \sum_i
    \mathbb{E}[X_i] \\
    &amp;= \sum_i e^{\mu_i + \frac{1}{2}\sigma_i^2}
    \\
    \sigma^2_{Z} &amp;= \frac{1}{S_+^2} \, \sum_{i,j}
      \rho_{ij} \sigma_i \sigma_j \mathbb{E}[X_i] \mathbb{E}[X_j] \\
      &amp;= \frac{1}{S_+^2} \, \sum_{i,j}
      \rho_{ij} \sigma_i \sigma_j e^{\mu_i+\frac{1}{2}\sigma_i^2}
      e^{\mu_j+\frac{1}{2}\sigma_j^2}
    \\
    \mu_Z &amp;= \log S_+ - \frac{1}{2}\sigma_{Z}^2
\end{aligned}
\tag{2}
\end{eqnarray}
</annotation></semantics></math></p>
<p>The sum term in the denominator of the renewal equation (1) consists
of a set of correlated scaled log normal distributions with scale
defined by the infectivity profile
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ï‰</mi><annotation encoding="application/x-tex">\omega</annotation></semantics></math>).
For our case for a given time point
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
we equate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>X</mi><mi>i</mi></msub><mo>=</mo><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><msub><mi>Î»</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub></mrow><annotation encoding="application/x-tex">X_i = \omega_{\tau} \lambda_{t-\tau}</annotation></semantics></math>,
and substitute
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¼</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mu_i</annotation></semantics></math>
=
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î¼</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub><mo>+</mo><mi>l</mi><mi>o</mi><mi>g</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mu_{t-\tau} + log(\omega_{\tau})</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Ïƒ</mi><mi>i</mi></msub><mo>=</mo><msub><mi>Ïƒ</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\sigma_i = \sigma_{t-\tau}</annotation></semantics></math>
into (2) to account for the infectivity profile. We define
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
to be the support of the infectivity profile
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">|</mo></mrow></mrow><annotation encoding="application/x-tex">k = |\omega|</annotation></semantics></math>).
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>m</mi><mi>s</mi></msub><annotation encoding="application/x-tex">m_{s}</annotation></semantics></math>
is the weighted contribution from incidence estimates on day
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow><annotation encoding="application/x-tex">t-\tau</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î£</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\Sigma_{ij}</annotation></semantics></math>
is the covariance between log-incidence estimates from days
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>âˆ’</mo><mi>i</mi></mrow><annotation encoding="application/x-tex">t-i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>âˆ’</mo><mi>j</mi></mrow><annotation encoding="application/x-tex">t-j</annotation></semantics></math>.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>m</mi><mi>Ï„</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mi>e</mi><mrow><msub><mi>Î¼</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub><mo>+</mo><mo>log</mo><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><mo>+</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow><mn>2</mn></msubsup></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î£</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mtext mathvariant="normal">Cov</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>Î»</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>i</mi></mrow></msub><mo>,</mo><mo>log</mo><msub><mi>Î»</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>Ï</mi><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>t</mi><mo>âˆ’</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>t</mi><mo>âˆ’</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msub><msub><mi>Ïƒ</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>i</mi></mrow></msub><msub><mi>Ïƒ</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>j</mi></mrow></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>S</mi><mo>+</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><munderover><mo>âˆ‘</mo><mrow><mi>Ï„</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><msub><mi>m</mi><mi>Ï„</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mrow><munderover><mo>âˆ‘</mo><mrow><mi>i</mi><mo>,</mo><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>m</mi><mi>i</mi></msub><msub><mi>m</mi><mi>j</mi></msub><msub><mi>Î£</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><msubsup><mi>S</mi><mo>+</mo><mn>2</mn></msubsup></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î¼</mi><mi>Z</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mo>log</mo><msub><mi>S</mi><mo>+</mo></msub><mo>âˆ’</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
    m_{\tau} &amp;= e^{\mu_{t-\tau} + \log \omega_{\tau} + \frac{1}{2}\sigma_{t-\tau}^2 } \\
    \Sigma_{ij} &amp;= \text{Cov}(\log \lambda_{t-i}, \log \lambda_{t-j}) = \rho_{(t-i)(t-j)} \sigma_{t-i} \sigma_{t-j} \\
    S_{+} &amp;= \sum_{\tau=1}^k { m_{\tau} } \\
    \sigma_{Z}^2 &amp;=
      \frac{
        \sum_{i,j=1}^k {
          (m_{i} m_{j} \Sigma_{ij})
        }
      }{S_{+}^2}    \\
    \mu_{Z} &amp;= \log S_+ - \frac{1}{2}\sigma_{Z}^2
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>With
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¼</mi><mi>Z</mi></msub><annotation encoding="application/x-tex">\mu_{Z}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>Z</mi></msub><annotation encoding="application/x-tex">\sigma_{Z}</annotation></semantics></math>
defined,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
is approximated as the ratio of two log-normals where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Z</mi></mrow></msub><mo>=</mo><mtext mathvariant="normal">Cov</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>Î»</mi><mi>t</mi></msub><mo>,</mo><mo>log</mo><mi>S</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\Sigma_{0Z} = \text{Cov}(\log \lambda_t, \log S)</annotation></semantics></math>
is the covariance between the numerator and the log-denominator. Since
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mo>=</mo><msub><mo>âˆ‘</mo><mi>Ï„</mi></msub><msub><mi>X</mi><mi>Ï„</mi></msub></mrow><annotation encoding="application/x-tex">S = \sum_\tau X_\tau</annotation></semantics></math>,
and using a first-order approximation, this covariance is a weighted
average of the covariances between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><msub><mi>Î»</mi><mi>t</mi></msub></mrow><annotation encoding="application/x-tex">\log \lambda_t</annotation></semantics></math>
and each
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><msub><mi>Î»</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\log \lambda_{t-\tau}</annotation></semantics></math>,
weighted by the relative expected contributions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>m</mi><mi>Ï„</mi></msub><annotation encoding="application/x-tex">m_\tau</annotation></semantics></math>.:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>R</mi><mi>t</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>âˆ¼</mo><mfrac><mrow><mi>L</mi><mi>o</mi><mi>g</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mi>a</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mi>t</mi></msub><mo>,</mo><msub><mi>Ïƒ</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><mi>L</mi><mi>o</mi><mi>g</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mi>a</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mi>Z</mi></msub><mo>,</mo><msub><mi>Ïƒ</mi><mi>Z</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î¼</mi><msub><mi>R</mi><mi>t</mi></msub></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msub><mi>Î¼</mi><mi>t</mi></msub><mo>âˆ’</mo><msub><mi>Î¼</mi><mi>Z</mi></msub><mo>=</mo><msub><mi>Î¼</mi><mi>t</mi></msub><mo>âˆ’</mo><mo>log</mo><msub><mi>S</mi><mo>+</mo></msub><mo>+</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Ïƒ</mi><msub><mi>R</mi><mi>t</mi></msub></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msqrt><mrow><msubsup><mi>Ïƒ</mi><mi>t</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup><mo>âˆ’</mo><mn>2</mn><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Z</mi></mrow></msub></mrow></msqrt></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Z</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mrow><munderover><mo>âˆ‘</mo><mrow><mi>Ï„</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><msub><mi>m</mi><mi>Ï„</mi></msub><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Ï„</mi></mrow></msub></mrow><mrow><munderover><mo>âˆ‘</mo><mrow><mi>Ï„</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><msub><mi>m</mi><mi>Ï„</mi></msub></mrow></mfrac><mo>=</mo><mfrac><mn>1</mn><msub><mi>S</mi><mo>+</mo></msub></mfrac><munderover><mo>âˆ‘</mo><mrow><mi>Ï„</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><msub><mi>m</mi><mi>Ï„</mi></msub><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Ï„</mi></mrow></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>R</mi><mi>t</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>âˆ¼</mo><mi>L</mi><mi>o</mi><mi>g</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mi>a</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><msub><mi>R</mi><mi>t</mi></msub></msub><mo>,</mo><msub><mi>Ïƒ</mi><msub><mi>R</mi><mi>t</mi></msub></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\tag{3}
\begin{aligned}
R_t &amp;\sim \frac{Lognormal(\mu_t,\sigma_t)}
{Lognormal( \mu_{Z}, \sigma_{Z})} \\
\mu_{R_t} &amp;= \mu_t - \mu_{Z} = \mu_t - \log S_+ + \frac{1}{2}\sigma_{Z}^2\\
\sigma_{R_t} &amp;= \sqrt{\sigma_t^2+\sigma_{Z}^2 - 2 \Sigma_{0Z}} \\
\Sigma_{0Z} &amp;= \frac{\sum_{\tau=1}^k m_\tau \Sigma_{0\tau}}{\sum_{\tau=1}^k m_\tau} = \frac{1}{S_+}\sum_{\tau=1}^k m_\tau \Sigma_{0\tau}\\
R_t &amp;\sim Lognormal(\mu_{R_t}, \sigma_{R_t})
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>This formulation of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
from (3) assumes knowledge of the posterior or prediction covariance of
the incidence estimates
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î£</mi><mi>i</mi></msub><mi>j</mi></mrow><annotation encoding="application/x-tex">\Sigma_ij</annotation></semantics></math>).
This is typical in modern frameworks , but in other situations may not
be available. In this case we could assume the individual estimates of
the incidence are independent, however this increases the uncertainty of
our
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
estimate and in certain circumstances introduces a potential
underestimation bias, influenced by the true off-diagonal correlation
mass and the certainty of the incidence estimates. An alternative
approach is to assume weak stationarity and estimate a parametric
correlation model from the data used to build the incidence model, using
Pearson residuals to parameterise an exponential decay function based on
time difference (see supporting software package for implementation).
The degree of the bias is investigated further in the supplementary
materials, and heuristics for assessing the significance of this bias
proposed.</p>
<p>The method for estimating
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
from modelled incidence has been described assuming a non-negative
component to the infectivity profile, as it is implicit that infector
and infectee are necessarily sequential in time. In the situation where
symptomatic case counts are used as a proxy for incidence and the serial
interval as a proxy for the infectivity profile, negative times between
serial cases may be observed due to variation in delay in observation of
the transmission chain. There is nothing in this framework to stop the
use of a negative time for the infectivity profile, and we can directly
support
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
estimates in these cases.</p>
<div class="section level3">
<h3 id="numerical-stability">Numerical stability<a class="anchor" aria-label="anchor" href="#numerical-stability"></a>
</h3>
<p>In (3)
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¼</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\mu_t</annotation></semantics></math>
is the log-scale mean of the incidence estimate at time
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\sigma_t</annotation></semantics></math>
its standard deviation. These can be large, leading to numerical
instability in terms involving
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>Î¼</mi><mo>+</mo><msup><mi>Ïƒ</mi><mn>2</mn></msup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\exp(\mu + \sigma^2)</annotation></semantics></math>.
However, assuming non-negative correlations and using log-space
computation with optimized log-sum-exp functions , the expressions
remain computationally tractable:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><mo>log</mo><msub><mi>m</mi><mi>Ï„</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msub><mi>Î¼</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow></msub><mo>+</mo><mo>log</mo><msub><mi>Ï‰</mi><mi>Ï„</mi></msub><mo>+</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mrow><mi>t</mi><mo>âˆ’</mo><mi>Ï„</mi></mrow><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mo>log</mo><msub><mi>S</mi><mo>+</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msub><mtext mathvariant="normal">logsumexp</mtext><mi>Ï„</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>m</mi><mi>Ï„</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mo>log</mo><msubsup><mi>Ïƒ</mi><mi>Z</mi><mn>2</mn></msubsup></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msub><mtext mathvariant="normal">logsumexp</mtext><mrow><mi>i</mi><mo>,</mo><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>m</mi><mi>i</mi></msub><mo>+</mo><mo>log</mo><msub><mi>m</mi><mi>j</mi></msub><mo>+</mo><mo>log</mo><msub><mi>Î£</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>âˆ’</mo><mn>2</mn><mo>log</mo><msub><mi>S</mi><mo>+</mo></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mo>log</mo><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Z</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msub><mtext mathvariant="normal">logsumexp</mtext><mi>Ï„</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>m</mi><mi>Ï„</mi></msub><mo>+</mo><mo>log</mo><msub><mi>Î£</mi><mrow><mn>0</mn><mi>Ï„</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>âˆ’</mo><mo>log</mo><msub><mi>S</mi><mo>+</mo></msub></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
\log m_\tau &amp;= \mu_{t-\tau} + \log \omega_\tau + \frac{1}{2} \sigma_{t-\tau}^2 \\
\log S_+ &amp;= \text{logsumexp}_\tau(\log m_\tau) \\
\log \sigma_Z^2 &amp;= \text{logsumexp}_{i,j} \left( \log m_i + \log m_j + \log \Sigma_{ij} \right) - 2 \log S_+ \\
\log \Sigma_{0Z} &amp;= \text{logsumexp}_\tau \left( \log m_\tau + \log \Sigma_{0\tau} \right) - \log S_+
\end{aligned}
\tag{4}
\end{eqnarray}
</annotation></semantics></math></p>
<p>Other relations may be implemented directly as in (3).</p>
</div>
<div class="section level3">
<h3 id="infectivity-profile-uncertainty">Infectivity profile uncertainty<a class="anchor" aria-label="anchor" href="#infectivity-profile-uncertainty"></a>
</h3>
<p>This estimate of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
is conditioned on a single known infectivity profile. In reality there
is also uncertainty in the infectivity profile
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Ï‰</mi><annotation encoding="application/x-tex">\omega</annotation></semantics></math>)
which plays a role in the definition of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¼</mi><mrow><mi>Z</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{Z,t}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mrow><mi>Z</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\sigma_{Z,t}</annotation></semantics></math>.
We cannot assume any particular distributional form for the infectivity
profile, but we can use a range of empirical estimates of the
infectivity profile to calculate multiple distributional estimates for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
and then combine these as a mixture distribution.</p>
<p>The nature of this mixture distribution will depend on the various
empirical infectivity profile distributions. However, we can use general
properties of mixture distributions to create estimates for the mean and
variance of the reproduction number estimate
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><annotation encoding="application/x-tex">R_t^*</annotation></semantics></math>)
combining the uncertainty arising from multiple infection profile
estimates
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Î©</mi><annotation encoding="application/x-tex">\Omega</annotation></semantics></math>)
and from the incidence estimate model itself:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>R</mi><mi>t</mi></msub><mo stretchy="false" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">]</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msup><mi>e</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub><mo>âˆ’</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><msubsup><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>V</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>R</mi><mi>t</mi></msub><mo stretchy="false" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">]</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">[</mo><msup><mi>e</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo>âˆ’</mo><mn>1</mn><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">]</mo><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">[</mo><msup><mi>e</mi><mrow><mn>2</mn><msub><mi>Î¼</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub><mo>+</mo><msubsup><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow><mn>2</mn></msubsup></mrow></msup><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">]</mo></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>R</mi><mi>t</mi></msub><mo stretchy="false" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">]</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>V</mi><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">[</mo><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mrow><mi>V</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>R</mi><mi>t</mi></msub><mo stretchy="false" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">]</mo></mrow><mo>+</mo><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>R</mi><mi>t</mi></msub><mo stretchy="false" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup></mrow><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">]</mo><mo>âˆ’</mo><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
E[R_t|\omega] &amp;= e^{(\mu_{R_t,\omega} - \frac{1}{2}\sigma_{R_t,\omega}^2)} \\
V[R_t|\omega] &amp;= \big[e^{(\sigma_{R_t,\omega}^2)} - 1\big] \big[e^{2 \mu_{R_t,\omega} + \sigma_{R_t,\omega}^2}\big] \\
E[R_t^*] &amp;= \frac{1}{|\Omega|}\sum_{\omega \in \Omega} E[{R_t|\omega}] \\
V[R_t^*] &amp;= \frac{1}{|\Omega|} \bigg[\sum_{\omega \in \Omega}{V[R_t|\omega]+E[R_t|\omega]^2}\bigg] - E[R_t^*]^2 \\
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>The cumulative distribution function of the mixture is simply the
arithmetic mean of the component cumulative distribution functions
(conditioned on each infectivity profile). If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Î¦</mi><annotation encoding="application/x-tex">\Phi</annotation></semantics></math>
is the cumulative distribution function of the standard normal
distribution:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>F</mi><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><msub><mi>F</mi><msub><mi>R</mi><mi>t</mi></msub></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="false" form="prefix">|</mo><mi>Ï‰</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo>â‰¤</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>R</mi><mrow><mi>t</mi><mo>,</mo><mi>Ï‰</mi></mrow></msub><mo>â‰¤</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo>â‰¤</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mi>Î¦</mi><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>âˆ’</mo><msub><mi>Î¼</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mrow><msub><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
F_{R_t^*}(x) &amp;= \frac{1}{|\Omega|}\sum_{\omega \in \Omega}F_{R_t}(x|\omega) \\
P(R_t^* \le x) &amp;= \frac{1}{|\Omega|}\sum_{\omega \in \Omega} P(R_{t,\omega} \le x) \\
P(R_t^* \le x) &amp;= \frac{1}{|\Omega|}\sum_{\omega \in \Omega} \Phi\bigg(\frac{ln(x) - \mu_{R_t,\omega}}{\sigma_{R_t,\omega}}\bigg)
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>As the cumulative density function of this mixture distribution is a
strictly increasing function, specific solutions for median
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>q</mi><mn>0.5</mn></msub><annotation encoding="application/x-tex">q_{0.5}</annotation></semantics></math>)
and 95% confidence intervals
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>q</mi><mn>0.025</mn></msub><annotation encoding="application/x-tex">q_{0.025}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>q</mi><mn>0.975</mn></msub><annotation encoding="application/x-tex">q_{0.975}</annotation></semantics></math>)
can be calculated numerically by solving the following equations:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mi>Î¦</mi><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>q</mi><mn>0.025</mn></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>âˆ’</mo><msub><mi>Î¼</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mrow><msub><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mo>âˆ’</mo><mn>0.025</mn></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mi>Î¦</mi><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>q</mi><mn>0.5</mn></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>âˆ’</mo><msub><mi>Î¼</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mrow><msub><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mo>âˆ’</mo><mn>0.5</mn></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mfrac><mn>1</mn><mrow><mo stretchy="true" form="prefix">|</mo><mi>Î©</mi><mo stretchy="true" form="postfix">|</mo></mrow></mfrac><munder><mo>âˆ‘</mo><mrow><mi>Ï‰</mi><mo>âˆˆ</mo><mi>Î©</mi></mrow></munder><mi>Î¦</mi><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>q</mi><mn>0.975</mn></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>âˆ’</mo><msub><mi>Î¼</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mrow><msub><mi>Ïƒ</mi><mrow><msub><mi>R</mi><mi>t</mi></msub><mo>,</mo><mi>Ï‰</mi></mrow></msub></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mo>âˆ’</mo><mn>0.975</mn></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>0</mn></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
\frac{1}{|\Omega|}\sum_{\omega \in \Omega} \Phi\bigg(\frac{ln(q_{0.025}) - \mu_{R_t,\omega}}{\sigma_{R_t,\omega}}\bigg) - 0.025 &amp;= 0 \\
\frac{1}{|\Omega|}\sum_{\omega \in \Omega} \Phi\bigg(\frac{ln(q_{0.5}) - \mu_{R_t,\omega}}{\sigma_{R_t,\omega}}\bigg) - 0.5 &amp;= 0 \\
\frac{1}{|\Omega|}\sum_{\omega \in \Omega} \Phi\bigg(\frac{ln(q_{0.975}) - \mu_{R_t,\omega}}{\sigma_{R_t,\omega}}\bigg) - 0.975 &amp;= 0
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>Numerical solutions to this are moderately expensive to perform. A
reasonable approximation can be expected by matching moments of a log
normal distribution to the mean
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">E[R_t^*]</annotation></semantics></math>
and variance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">V[R_t^*]</annotation></semantics></math>
of the mixture. This gives us the final closed form estimator for the
reproduction number given a set of infectivity profiles,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><msub><mi>R</mi><mrow><mi>t</mi><mo>,</mo><mi>Î©</mi></mrow></msub><mo accent="true">Â¯</mo></mover><annotation encoding="application/x-tex">\overline{R_{t,\Omega}}</annotation></semantics></math>,
as:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Î¼</mi><mrow><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>Î©</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>l</mi><mi>o</mi><mi>g</mi><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup></mrow><msqrt><mrow><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup><mo>+</mo><mi>V</mi><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow></mrow></msqrt></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Ïƒ</mi><mrow><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>Î©</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msqrt><mrow><mi>l</mi><mi>o</mi><mi>g</mi><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mn>1</mn><mo>+</mo><mfrac><mrow><mi>V</mi><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow></mrow><mrow><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>R</mi><mi>t</mi><mo>*</mo></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup></mrow></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo></mrow></msqrt></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mover><msub><mi>R</mi><mrow><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>Î©</mi></mrow></msub><mo accent="true">Â¯</mo></mover></mtd><mtd columnalign="left" style="text-align: left"><mo>âˆ¼</mo><mi>L</mi><mi>o</mi><mi>g</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mi>a</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î¼</mi><mrow><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>Î©</mi></mrow></msub><mo>,</mo><msub><mi>Ïƒ</mi><mrow><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>Î©</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{eqnarray}
\begin{aligned}
\mu_{t|\Omega} &amp;= log\bigg(\frac{E[R_t^*]^2}{\sqrt{E[R_t^*]^2 + V[R_t^*]}}\bigg) \\
\sigma_{t|\Omega} &amp;= \sqrt{log\bigg(1 + \frac{V[R_t^*]}{E[R_t^*]^2}\bigg)}\\
\overline{R_{t|\Omega}} &amp;\sim Lognormal(\mu_{t|\Omega},\sigma_{t|\Omega})
\end{aligned}
\end{eqnarray}
</annotation></semantics></math></p>
<p>In summary we present a method for retrieving the distributional form
of the reproduction number from log normally distributed probabilistic
estimates of incidence arising from simple statistical count models.
This includes uncertainty arising from both count models and from
infectivity profile distributions. It is fully deterministic and
computationally inexpensive. It does not place any particular
constraints on the nature of the infectivity profile distribution and
can handle distributions that have a negative component, as is sometimes
seen in serial interval estimates.</p>
</div>
</div>
<div class="section level2">
<h2 id="implementation">Implementation<a class="anchor" aria-label="anchor" href="#implementation"></a>
</h2>
<p>This method is implemented using the following R function, which is
designed for numerical stability and speed. Generating
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
estimates given modelled incidence typically occurring in a matter of
seconds:</p>
<pre><code><span><span class="co">#&gt; function (mu_t, vcov_ij = diag(sigma_t^2), omega, sigma_t = NULL, </span></span>
<span><span class="co">#&gt;     tau_offset = 0) </span></span>
<span><span class="co">#&gt; {</span></span>
<span><span class="co">#&gt;     omega_m = as.matrix(omega)</span></span>
<span><span class="co">#&gt;     k = nrow(omega_m)</span></span>
<span><span class="co">#&gt;     if (k != length(mu_t)) {</span></span>
<span><span class="co">#&gt;         stop("omega must have the same number of rows as length of mu_t.")</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     if (!all(dim(vcov_ij) == k)) {</span></span>
<span><span class="co">#&gt;         stop("vcov_ij must be a square matrix with the same dimensions as mu_t, and nrow(omega)")</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     omega_m = apply(omega_m, MARGIN = 2, rev)</span></span>
<span><span class="co">#&gt;     sigma_t = sqrt(diag(vcov_ij))</span></span>
<span><span class="co">#&gt;     time_Rt = length(mu_t) - tau_offset</span></span>
<span><span class="co">#&gt;     mu = mu_t[time_Rt]</span></span>
<span><span class="co">#&gt;     sigma = sigma_t[time_Rt]</span></span>
<span><span class="co">#&gt;     tmp = apply(omega_m, MARGIN = 2, function(omega) {</span></span>
<span><span class="co">#&gt;         log_m_tau = mu_t + sigma_t^2/2 + log(omega)</span></span>
<span><span class="co">#&gt;         log_S_plus = .logsumexp(log_m_tau)</span></span>
<span><span class="co">#&gt;         log_Sigma_ij = log(pmax(vcov_ij, 0))</span></span>
<span><span class="co">#&gt;         log_Sigma_0tau = log_Sigma_ij[time_Rt, ]</span></span>
<span><span class="co">#&gt;         log_sigma_Z2 = .logsumexp(outer(log_m_tau, log_m_tau, </span></span>
<span><span class="co">#&gt;             "+") + log_Sigma_ij) - 2 * log_S_plus</span></span>
<span><span class="co">#&gt;         mu_Z = log_S_plus - 1/2 * exp(log_sigma_Z2)</span></span>
<span><span class="co">#&gt;         log_Sigma_0Z = .logsumexp(log_m_tau + log_Sigma_0tau) - </span></span>
<span><span class="co">#&gt;             log_S_plus</span></span>
<span><span class="co">#&gt;         mu_Rt = mu - mu_Z</span></span>
<span><span class="co">#&gt;         sigma_Rt2 = sigma^2 + exp(log_sigma_Z2) - 2 * exp(log_Sigma_0Z)</span></span>
<span><span class="co">#&gt;         return(c(mu_Rt, sigma_Rt2))</span></span>
<span><span class="co">#&gt;     })</span></span>
<span><span class="co">#&gt;     mu_Rt = tmp[1, ]</span></span>
<span><span class="co">#&gt;     sigma_Rt2 = tmp[2, ]</span></span>
<span><span class="co">#&gt;     means = exp(mu_Rt + sigma_Rt2/2)</span></span>
<span><span class="co">#&gt;     vars = (exp(sigma_Rt2) - 1) * exp(2 * mu_Rt + sigma_Rt2)</span></span>
<span><span class="co">#&gt;     mean_star = mean(means)</span></span>
<span><span class="co">#&gt;     var_star = mean(vars + means^2) - mean_star^2</span></span>
<span><span class="co">#&gt;     mu_star = log(mean_star) - log(var_star/mean_star^2 + 1)/2</span></span>
<span><span class="co">#&gt;     sigma_star = sqrt(log(1 + var_star/mean_star^2))</span></span>
<span><span class="co">#&gt;     return(list(time_Rt = time_Rt, mean_Rt_star = mean_star, </span></span>
<span><span class="co">#&gt;         var_Rt_star = var_star, meanlog_Rt_star = mu_star, sdlog_Rt_star = sigma_star, </span></span>
<span><span class="co">#&gt;         mu_Rt_mix = mu_Rt, sigma_Rt_mix = sqrt(sigma_Rt2), quantile_Rt_fn = function(p) {</span></span>
<span><span class="co">#&gt;             .qmixlnorm(p, mu_Rt, sqrt(sigma_Rt2))</span></span>
<span><span class="co">#&gt;         }))</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; &lt;bytecode: 0x63c5123313d8&gt;</span></span>
<span><span class="co">#&gt; &lt;environment: namespace:ggoutbreak&gt;</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>Testing this against the incidence model shown above, and comparing
the results to the SPI-M-O consensus
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
estimates gives us the following time-series for England. This is
formally evaluated elsewhere but qualitatively is a good fit. It took a
few seconds to calculate the reproduction number from this single time
series with 1410 time points , which opens up the possibility of
performing
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
estimates in fine grained geographical or demographic subgroups.</p>
<pre><code><span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  10.879   1.146  12.030</span></span></code></pre>
<p><img src="rt-from-incidence_files/figure-html/unnamed-chunk-3-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>We present a methodology for deriving
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><mi>t</mi></msub><annotation encoding="application/x-tex">R_t</annotation></semantics></math>
from modelled estimates of incidence while propagating uncertainty. We
demonstrate it produces satisfactory qualitative results against
COVID-19 data. This method is relatively quick, fully deterministic, and
can be used on top of any statistical models for estimating incidence
which use logarithmic link functions.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Robert Challen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
