<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multinomial proportions models for genomic variants • ggoutbreak</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multinomial proportions models for genomic variants">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggoutbreak</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ggoutbreak.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/covid-timeseries.html">England COVID-19 cases</a></li>
    <li><a class="dropdown-item" href="../articles/estimators-example.html">Simulation tests for growth rate estimators</a></li>
    <li><a class="dropdown-item" href="../articles/incidence-trends.html">Population comparisons and incidence</a></li>
    <li><a class="dropdown-item" href="../articles/infectivity-profile-discretisation.html">Infectivity profile discretisation</a></li>
    <li><a class="dropdown-item" href="../articles/right-censoring.html">Estimating the reproduction number from right censored data</a></li>
    <li><a class="dropdown-item" href="../articles/rt-from-incidence.html">Estimating the reproduction number from modelled incidence</a></li>
    <li><a class="dropdown-item" href="../articles/sampling-serial-interval.html">Sampling the infectivity profile from published serial interval estimates</a></li>
    <li><a class="dropdown-item" href="../articles/simulation-test-models.html">Simulations and test harnesses</a></li>
    <li><a class="dropdown-item" href="../articles/time-periods.html">Data wrangling and working with `ggoutbreak`</a></li>
    <li><a class="dropdown-item" href="../articles/variant-proportions.html">Multinomial proportions models for genomic variants</a></li>
    <li><a class="dropdown-item" href="../articles/weekly-incidence.html">Estimating the reproduction number from weekly data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ai4ci/ggoutbreak/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multinomial proportions models for genomic variants</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ai4ci/ggoutbreak/blob/HEAD/vignettes/variant-proportions.Rmd" class="external-link"><code>vignettes/variant-proportions.Rmd</code></a></small>
      <div class="d-none name"><code>variant-proportions.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="covid-19-proportions-in-england">COVID-19 proportions in England<a class="anchor" aria-label="anchor" href="#covid-19-proportions-in-england"></a>
</h2>
<p>The Sanger Centre &amp; COGUK performed a large amount of sequencing
of COVID-19 during the pandemic, to identify emerging genomic variants.
This was scaled up in the second half of 2021 and continued through to
the beginning of 2023. Lineages were assigned using the Pango lineage
system and important ones given nicknames by the WHO.</p>
<p>The Sanger variants data has been discontinued, but were still
available for download. The code to download, process these data sets
and determine the full lineage is in the
<code>data-raw/variants.R</code> file, but the output of this has been
bundled as a data set here. There are many caveats to the data here in
terms of bias and it should not be regarded as definitive:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># tidy copy of the sanger weekly variants count data aggregated to England level</span></span>
<span><span class="fu">ggoutbreak</span><span class="fu">::</span><span class="va"><a href="../reference/england_variants.html">england_variants</a></span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 479</span></span>
<span><span class="co">#&gt; Columns: 6</span></span>
<span><span class="co">#&gt; Groups: class [10]</span></span>
<span><span class="co">#&gt; $ date      <span style="color: #949494; font-style: italic;">&lt;date&gt;</span> 2020-09-05<span style="color: #949494;">, </span>2020-09-05<span style="color: #949494;">, </span>2020-09-12<span style="color: #949494;">, </span>2020-09-12<span style="color: #949494;">, </span>2020-09-19<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ time      <span style="color: #949494; font-style: italic;">&lt;time_prd&gt;</span> 0<span style="color: #949494;">, </span>0<span style="color: #949494;">, </span>7<span style="color: #949494;">, </span>7<span style="color: #949494;">, </span>14<span style="color: #949494;">, </span>14<span style="color: #949494;">, </span>21<span style="color: #949494;">, </span>21<span style="color: #949494;">, </span>28<span style="color: #949494;">, </span>28<span style="color: #949494;">, </span>35<span style="color: #949494;">, </span>35<span style="color: #949494;">, </span>42<span style="color: #949494;">, </span>42<span style="color: #949494;">, </span>49<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ class     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Other<span style="color: #949494;">, </span>Alpha (B.1.1.7)<span style="color: #949494;">, </span>Other<span style="color: #949494;">, </span>Alpha (B.1.1.7)<span style="color: #949494;">, </span>Other<span style="color: #949494;">, </span>Alpha…</span></span>
<span><span class="co">#&gt; $ who_class <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Other<span style="color: #949494;">, </span>Alpha<span style="color: #949494;">, </span>Other<span style="color: #949494;">, </span>Alpha<span style="color: #949494;">, </span>Other<span style="color: #949494;">, </span>Alpha<span style="color: #949494;">, </span>Other<span style="color: #949494;">, </span>Alpha<span style="color: #949494;">, </span>Othe…</span></span>
<span><span class="co">#&gt; $ count     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 1182<span style="color: #949494;">, </span>371<span style="color: #949494;">, </span>1439<span style="color: #949494;">, </span>588<span style="color: #949494;">, </span>837<span style="color: #949494;">, </span>429<span style="color: #949494;">, </span>1685<span style="color: #949494;">, </span>1157<span style="color: #949494;">, </span>1208<span style="color: #949494;">, </span>823<span style="color: #949494;">, </span>1501<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ denom     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 1553<span style="color: #949494;">, </span>1553<span style="color: #949494;">, </span>2027<span style="color: #949494;">, </span>2027<span style="color: #949494;">, </span>1266<span style="color: #949494;">, </span>1266<span style="color: #949494;">, </span>2842<span style="color: #949494;">, </span>2842<span style="color: #949494;">, </span>2031<span style="color: #949494;">, </span>2031<span style="color: #949494;">, </span>…</span></span></code></pre></div>
<p>The data must have a <code>class</code> column defining the main
categorisation of the data (in this case it is the main Pango variant).
The <code>time</code> column is a <code>time_period</code> derived from
the date (which is weekly). The other necessary column is the
<code>count</code> column which is integer counts of each
<code>class</code>. The data must be grouped by <code>class</code>.
Multiple models can be fitted simultaneously if the data is grouped by
other columns.</p>
</div>
<div class="section level2">
<h2 id="multinomial-proportions-model-">Multinomial proportions model.<a class="anchor" aria-label="anchor" href="#multinomial-proportions-model-"></a>
</h2>
<p>Genomic testing happened only in a subset of cases. The testing
effort varied significantly over time. The proportion of each variant
over time can be estimated with a multinomial model.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">probs</span> <span class="op">=</span> <span class="va">england_variants</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/multinomial_nnet_model.html">multinomial_nnet_model</a></span><span class="op">(</span>window <span class="op">=</span> <span class="fl">28</span><span class="op">)</span></span>
<span><span class="co">#&gt; # weights:  40 (27 variable)</span></span>
<span><span class="co">#&gt; initial  value 3583520.087982 </span></span>
<span><span class="co">#&gt; iter  10 value 1562162.620103</span></span>
<span><span class="co">#&gt; iter  20 value 1380207.974678</span></span>
<span><span class="co">#&gt; iter  30 value 959088.555894</span></span>
<span><span class="co">#&gt; iter  40 value 742507.401941</span></span>
<span><span class="co">#&gt; iter  50 value 731767.878774</span></span>
<span><span class="co">#&gt; iter  60 value 729647.595794</span></span>
<span><span class="co">#&gt; iter  70 value 726795.599521</span></span>
<span><span class="co">#&gt; iter  80 value 716309.104988</span></span>
<span><span class="co">#&gt; iter  90 value 709310.965531</span></span>
<span><span class="co">#&gt; iter 100 value 707285.575109</span></span>
<span><span class="co">#&gt; final  value 707285.575109 </span></span>
<span><span class="co">#&gt; stopped after 100 iterations</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_multinomial.html">plot_multinomial</a></span><span class="op">(</span><span class="va">probs</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_d</a></span><span class="op">(</span>option<span class="op">=</span><span class="st">"cividis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="variant-proportions_files/figure-html/unnamed-chunk-2-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="binomial-proportions-model">Binomial proportions model<a class="anchor" aria-label="anchor" href="#binomial-proportions-model"></a>
</h2>
<p>The binomial (one versus others) proportions are not very different
to the multinomial probabilities calculated above, but come with
confidence intervals, however the median values do not necessarily sum
to 1.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">probs2</span> <span class="op">=</span> <span class="va">england_variants</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/proportion_locfit_model.html">proportion_locfit_model</a></span><span class="op">(</span>window <span class="op">=</span> <span class="fl">14</span>, deg<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_proportion.html">plot_proportion</a></span><span class="op">(</span><span class="va">probs2</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_d</a></span><span class="op">(</span>option<span class="op">=</span><span class="st">"cividis"</span>,aesthetics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"colour"</span>,<span class="st">"fill"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="variant-proportions_files/figure-html/unnamed-chunk-3-1.png" width="576"></p>
<p>The rate of change of the proportion of each individual variant
versus the others on a logistic scale can be used to work out the
exponential growth rate of one variant relative to the others. Because
this is a relative growth rate taken together the estimates of all
variants at a given time are centred around zero. If one variant has a
growth advantage, by definition others have a growth disadvantage
despite potentially causing a larger disease burden and potentially
having increasing numbers in a growing epidemic.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_growth_rate.html">plot_growth_rate</a></span><span class="op">(</span><span class="va">probs2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_d</a></span><span class="op">(</span>option<span class="op">=</span><span class="st">"cividis"</span>,aesthetics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"colour"</span>,<span class="st">"fill"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="variant-proportions_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
<p>The binomial relative growth rate per day is a growth advantage over
existing variants. This has a dependency on the unit of time which is
controlled by the <code>time_period</code> configuration. In the data
provided here the <code>time_period</code> is defined on a daily basis
despite the data being provided weekly. Doubling time does not make
strict sense when describing relative growth rates and is not shown
here. The variant advantage in terms of reproduction number needs
additional information about the infectivity profile (aka generation
time distribution) and as a result is much more complex.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Robert Challen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
